### ISSUE TITLE
Horrendous white-space bug keeps Orbot from loading bridges

### ISSUE THREAD

**Issue by ghost** [guardianproject/orbot#289]
Hello
There is this bug that has been living happily in Orbot's code since it's earliest version and it is really, really surprising how it has evaded scrutiny right until the latest version.

Here's the short description: Orbot is very, very fragile when it comes to handling whitespaces that might exist in the bridge list fed into the "Bridges" Input box.

I'm almost certain that not a single one of Orbot's devs have *ever* tried using bridges **via email** -- the process is easy: 
  * log into your email, 
  * send "get transport obfs4" to "bridges@torproject.org"
  * when you receive the bridge list, Copy them all at one,
  * in Orbot, choose Settings -> Bridges, Paste the bridge list into the input box
  * restart Orbot, and watch as it **FAILS** to load the torrc file due to some unrecognised character!

I wish it was just that!

Orbot's "torrc" would also fail to parse in these conditions too:
  * when there is any whitespace after the '\n' separating the bridges; e.g:
  this would successfully parse:
  ```
obfs4 38.229.33.152:50614 AEDC774CF21732E86F233869469CD8E5EEFC7765 cert=pcaZ00YFYW72mrFAcXQAEzMJNxV+RhYMIDJHhhoGWdRzlQqKm3JE1xBR+Uis6bGqbAPaDQ iat-mode=1
obfs4 38.229.33.139:42141 4047D875CC00F1CDDFD432814AB4A2723BFC412D cert=b7ejF+izmtLL/fuMo0IMIjGCPmeUfB9B3uat/KyHGRldNWcoLZ50VNhcMZoETy0iHUfKSQ iat-mode=1
obfs4 13.81.48.139:43981 C128E91BCF541FE0F6323F51D6B3C67326E441F4 cert=GlYLwHazlIfTYpJETZRkjx2PNlcJAKajCV6CoqyR8/Xceph5EpstUK176KvTBJmaG+UnFA iat-mode=0
```

   ....... but this won't:
   ```
obfs4 38.229.33.152:50614 AEDC774CF21732E86F233869469CD8E5EEFC7765 cert=pcaZ00YFYW72mrFAcXQAEzMJNxV+RhYMIDJHhhoGWdRzlQqKm3JE1xBR+Uis6bGqbAPaDQ iat-mode=1

obfs4 38.229.33.139:42141 4047D875CC00F1CDDFD432814AB4A2723BFC412D cert=b7ejF+izmtLL/fuMo0IMIjGCPmeUfB9B3uat/KyHGRldNWcoLZ50VNhcMZoETy0iHUfKSQ iat-mode=1

obfs4 13.81.48.139:43981 C128E91BCF541FE0F6323F51D6B3C67326E441F4 cert=GlYLwHazlIfTYpJETZRkjx2PNlcJAKajCV6CoqyR8/Xceph5EpstUK176KvTBJmaG+UnFA iat-mode=0
   ```

The bottom line is, even though the software you have built is 99% excellent, this seemingly trivial 1%-bug has made it *totally* unusable for the grass-roots in countries like mine (Iran), where the Internet is totally meaningless in the absence of Tor-over-Bridges.

I am a developer myself, and nuisances like this are something I can find a way around to most of the time, but the majority of Orbot's users are mere users. They *need* something that doesn't break on things this trivial.

I wish the developers tried to use the Internet through Orbot themselves.
I wish they tried getting bridges through Gmail app.

Only then would they have found out how PAINFUL the process is.

Thanks for reading this far.

**Comment by n8fr8** [comment]
We had implemented code in the past to attempt to clean the whitespaces out, and ensure the newlines were properly there. I can find it if you'd like.  We have tested and used this feature, and do care about users in Iran, China and elsewhere that require bridges. Most of our energy goes into ensuring the bridge code runs in the first place, and can work if you do have a reachable bridge.

Still, we can do better. I'll take a look again how we handle these cases. Maybe we should tell users to not paste in the entirely three bridges, and just one line at a time.

The better option is to support the automated fetching of new bridges through a built in user interface like Tor Browser on the desktop does. We hope that can be implemented soon.

**Comment by bitmold** [comment]
I agree that implementing automated bridge fetching would be the right thing to do, but in the meantime I'm going to fix this up. Besides, even if we have automated bridge fetching we would still want to support pasting in bridges from the bridge email service (or from wherever else).

**Comment by bitmold** [comment]
So it seems like there are at least two parsing bugs in `OrbotService`'s `private StringBuffer processSettingsImpl (StringBuffer extraLines)`which is responsible for pulling out settings and dumping them into a torrc file at runtime. 

This is the relevant code:
``` 
String[] bridgeListLines = bridgeList.trim().split("\\n");
int bridgeIdx = (int)Math.round(Math.random()*((double)bridgeListLines.length));
String bridgeLine = bridgeListLines[bridgeIdx];
extraLines.append("Bridge ");
extraLines.append(bridgeLine);
extraLines.append("\n");
```   

The first issue is that when you paste this example in from this bug report 
```
obfs4 38.229.33.152:50614 AEDC774CF21732E86F233869469CD8E5EEFC7765 cert=pcaZ00YFYW72mrFAcXQAEzMJNxV+RhYMIDJHhhoGWdRzlQqKm3JE1xBR+Uis6bGqbAPaDQ iat-mode=1

obfs4 38.229.33.139:42141 4047D875CC00F1CDDFD432814AB4A2723BFC412D cert=b7ejF+izmtLL/fuMo0IMIjGCPmeUfB9B3uat/KyHGRldNWcoLZ50VNhcMZoETy0iHUfKSQ iat-mode=1

obfs4 13.81.48.139:43981 C128E91BCF541FE0F6323F51D6B3C67326E441F4 cert=GlYLwHazlIfTYpJETZRkjx2PNlcJAKajCV6CoqyR8/Xceph5EpstUK176KvTBJmaG+UnFA iat-mode=0
```

the array declared `bridgeListLines` will contain valid bridge lines but also newlines. Here's an example from the debugger at runtime:
![Screenshot from 2020-04-08 19-40-54](https://user-images.githubusercontent.com/22125581/78844693-7d462c80-79d4-11ea-8c3e-bf06831bb23a.png)

I'm still looking into this, but it seems that it there is a decent probability that the randomly selected bridge will be an empty string instead of a bridge. 

The second issue is a possible index out of bounds exception that could be thrown. The offending line is here: `int bridgeIdx = (int)Math.round(Math.random()*((double)bridgeListLines.length));`
By using `Math.round()` instead of `Math.floor()` you get values ranging from 0 to 5 (inclusive) instead of 0 to 5 exclusively. Hopefully changing these two things prevents these bug from happening...

**Comment by bitmold** [comment]
This can be closed since #311 has been merged

**Comment by n8fr8** [comment]
https://github.com/guardianproject/orbot/releases/tag/16.2.0-RC-1-tor-0.4.2.7
